programmers/level1/[1차]-다트-게임/김민준.js

```diff
@@ -0,0 +1,36 @@
+function solution(dartResult) {
+  let answer = 0;
+  let [prevScore, score] = [0, 0];
+
+  for (let i = 0; i < dartResult.length; i += 1) {
+    let dart = dartResult[i];
+
+    if ('0' <= dart && dart <= '9') {
+      prevScore = score;
+      // 점수가 10점일때
+      if (dartResult[i + 1] === '0') {
+        score = 10;
+        i += 1;
+      } else score = parseInt(dart);
+    } else if (dart === 'S' || 'D' || 'T') {
+      // 'S'는 생략 가능
+      if (dart === 'S') score = score ** 1;
+      else if (dart === 'D') score = score ** 2;
+      else if (dart === 'T') score = score ** 3;
+
+      if (dartResult[i + 1] === '*') {
+        // 이미 answer에 더해져있기 때문에 한번만 더함
+        answer += prevScore;
+        score *= 2;
+        i += 1;
+      } else if (dartResult[i + 1] === '#') {
+        score *= -1;
+        i += 1;
+      }
+
+      answer += score;
+    }
+  }
+
+  return answer;
+}
```

Comment:
리팩토링 해봤습니다. 정규표현식을 사용한 제 코드보다 성능이 두 배는 좋네요.

```suggestion
function solution(dartResult) {
  const BONUS = { S: 1, D: 2, T: 3 };
  const BONUSES = Object.keys(BONUS);
  const OPTION = { "#": -1, "*": 2 };
  const OPTIONS = Object.keys(OPTION);

  let answer = 0;
  let [prevScore, score] = [0, 0];

  for (let i = 0; i < dartResult.length; i += 1) {
    const dart = dartResult[i];

    if ("0" <= dart && dart <= "9") {
      prevScore = score;
      if (dartResult[i + 1] === "0") {
        score = 10;
        i += 1;
      } else score = parseInt(dart);
    }

    if (BONUSES.includes(dart)) {
      score = score ** BONUS[dart];

      const option = dartResult[i + 1];
      if (OPTIONS.includes(option)) {
        if (option === "*") answer += prevScore;
        score *= OPTION[option];
        i += 1;
      }

      answer += score;
    }
  }

  return answer;
}
```

Reply:
깔끔하네요! 참고하겠습니다